src/meshapi/util/join_records.py

```diff
@@ -0,0 +1,106 @@
+import datetime
+import json
+import logging
+import os
+import tempfile
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, fields
+from typing import Optional
+
+import boto3
+from botocore.client import ClientError, Config
+
+from meshapi.views.forms import JoinFormRequest
+
+JOIN_RECORD_ENDPOINT = os.environ.get("JOIN_RECORD_ENDPOINT")
+JOIN_RECORD_BUCKET_NAME = os.environ.get("JOIN_RECORD_BUCKET_NAME")
+JOIN_RECORD_BASE_NAME = os.environ.get("JOIN_RECORD_BASE_NAME", "sample-basename")
+JOIN_RECORD_ACCESS_KEY = os.environ.get("JOIN_RECORD_ACCESS_KEY")
+JOIN_RECORD_SECRET_KEY = os.environ.get("JOIN_RECORD_SECRET_KEY")
+
+
+@dataclass
+class JoinRecord(JoinFormRequest):
+    submission_time: str
+    code: str
+    replayed: int
+    install_number: Optional[int]
+
+
+def s3_content_to_join_record(object_key: str, content: str) -> JoinRecord:
+    content_json = json.loads(content)
+    mapped_content_json = {key.name: content_json.get(key.name) for key in fields(JoinRecord)}
+    join_record = JoinRecord(**mapped_content_json)
+
+    # Convert S3 path to datetime
+    # "dev-join-form-submissions/2024/10/28/12/27/00.json"
+    datetime_components = object_key.split(".")[0].split("/")[1:]
+    year, month, day, hour, minute, second = map(int, datetime_components)
+    result_datetime = datetime.datetime(year, month, day, hour, minute, second)
+    join_record.submission_time = result_datetime.isoformat()
+    return join_record
+
+
+class JoinRecordProcessorInterface(ABC):
+    @abstractmethod
+    def __init__(self) -> None:
+        pass
+
+    @abstractmethod
+    def upload(self, join_record: JoinRecord, key: str) -> None:
+        pass
+
+    @abstractmethod
+    def get_all(self) -> list[JoinRecord]:
+        pass
+
+
+class JoinRecordProcessor(JoinRecordProcessorInterface):
+    def __init__(self) -> None:
+        self.s3_client = boto3.client(
+            "s3",
+            endpoint_url=JOIN_RECORD_ENDPOINT,
```

Comment:
If this env var isn't provided, we should probably make this `None` so we let boto grab the default, it's kinda annoying to have to set this manually for real S3